src/consume/mod.rs

```diff
@@ -0,0 +1,202 @@
+        self.ack_token.ack().await?;
+        Ok(self.message)
+    }
+
+    /// Negatively acknowledge this message, declaring that processing was unsuccessful and the
+    /// message should be re-delivered to consumers.
+    pub async fn nack(self) -> Result<M, A::NackError> {
+        self.ack_token.nack().await?;
+        Ok(self.message)
+    }
+
+    /// Modify the acknowledgement deadline for this message to the given number of seconds.
+    ///
+    /// The new deadline will typically be this number of seconds after the service receives this
+    /// modification requesst, though users should check their implementation's documented
+    /// behavior.
+    pub async fn modify_deadline(&mut self, seconds: u32) -> Result<(), A::ModifyError> {
+        self.ack_token.modify_deadline(seconds).await
+    }
+}
+
+impl<A, M> std::ops::Deref for AcknowledgeableMessage<A, M> {
+    type Target = M;
+
+    fn deref(&self) -> &M {
+        &self.message
+    }
+}
+
+impl<A, M> std::ops::DerefMut for AcknowledgeableMessage<A, M> {
+    fn deref_mut(&mut self) -> &mut M {
+        &mut self.message
+    }
+}
+
+/// A token associated with some message received from a message service, used to issue an
+/// ack/nack/modify request
+///
+/// See the documentation for acknowledging messages on [`Consumer`]
+#[async_trait]
+#[must_use = "Messages should be ack'ed to prevent repeated delivery, or nack'ed to improve responsiveness"]
+pub trait AcknowledgeToken {
+    /// Errors returned by [`ack`](AcknowledgeToken::ack)
+    type AckError;
+    /// Errors returned by [`nack`](AcknowledgeToken::nack)
+    type NackError;
+    /// Errors returned by [`modify_deadline`](AcknowledgeToken::modify_deadline)
+    type ModifyError;
+
+    /// Acknowledge the associated message
+    async fn ack(self) -> Result<(), Self::AckError>;
+
+    /// Negatively acknowledge the associated message
+    async fn nack(self) -> Result<(), Self::NackError>;
+
+    /// Change the associated message's acknowledge deadline to the given number of seconds
+    // uses u32 seconds instead of e.g. Duration because SQS and PubSub both have second
+    // granularity; Duration::from_millis(999) would truncate to 0, which might be surprising
+    async fn modify_deadline(&mut self, seconds: u32) -> Result<(), Self::ModifyError>;
+}
+
+/// The stream returned by the [`consume`](Consumer::consume) function
+#[pin_project]
+pub struct MessageStream<S, V, M> {
+    #[pin]
+    stream: S,
+    validator: V,
+    _message_type: std::marker::PhantomData<M>,
+}
+
+impl<S, V, M, AckToken, StreamError> stream::Stream for MessageStream<S, V, M>
+where
+    S: stream::Stream<
+        Item = Result<AcknowledgeableMessage<AckToken, ValidatedMessage>, StreamError>,
+    >,
+    M: DecodableMessage<Validator = V>,
+{
+    #[allow(clippy::type_complexity)] // it is what it is, aliases would all be generic anyway
+    type Item = Result<AcknowledgeableMessage<AckToken, M>, Either<StreamError, M::Error>>;
```

Comment:
why not just create a wrapper error type instead of use Either? if will read better

Reply:
will it read better? Using Either felt like saving boilerplate compared to making a wrapper enum, but I can do that if it's clearer

Reply:
i guess if it does, not by a lot. probably fine to leave it